### DISCUSSION TITLE
DP+LAMMPS Precision

### DISCUSSION THREAD

**Discussion by npiroozan** [deepmodeling/deepmd-kit#1944] [Q&A]
Hello,

I know this has been discussed before but the solution is a bit unclear to me.  The following is the issue:

tensorflow/core/framework/tensor.cc:725] Check failed: dtype() == expected_dtype (1 vs. 2) double expected, got float

Clearly the model has been trained for single precision but LAMMPS is expecting inference output of double precision.  I'm using version 2.1.2 of DP-kit and build DP+LAMMPS with the plugin mode.

How can this be resolved?

Thank you very much!

Warm Regards

**Comment by njzjz** [comment]
See #1907, if you installed `lammps` from our conda channel instead of the old `lammps-dp` package.

**Reply by npiroozan** [reply]
Sorry for the bother, but what would need to be changed if using LAMMPS-dp?

In deepmdroot/bin I only have lmp, but no lmp_low.

Thank you!

**Reply by njzjz** [reply]
Backup `lib/libdeepmd_lmp.so` and move `lib/libdeepmd_lmp_low.so` to `lib/libdeepmd_lmp.so`.

**Reply by npiroozan** [reply]
It seems I don't have these dynamic library files.  I installed LAMMPS and DeePMD as a plugin based on the configurations specified in the manual.

**Reply by njzjz** [reply]
Did you build from source? If so, did you specify the cmake argument `LAMMPS_SOURCE_ROOT`?

**Reply by npiroozan** [reply]
Yes I built from source, but did not specify LAMMPS_SOURCE_ROOT.  It did not appear in the directives for building as a plug-in.

**Reply by njzjz** [reply]
![image](https://user-images.githubusercontent.com/9496702/194160603-3dc91fcd-451b-47c9-b982-86f9e9740528.png)

**Reply by npiroozan** [reply]
I deeply apologize, it seems I have forgotten an important step in this process.  That is namely, buidling the Tensorflow C++ interface.  Surprisingly, DP+LMP worked fine for me but the files created after cmake were incomplete, as noted earlier.

So it seems the following needs to be done first:

cmake -DTENSORFLOW_ROOT=$tensorflow_root -DCMAKE_INSTALL_PREFIX=$deepmd_root ..


Unfortunately, I have traditionally had serious problems doing this in the past, specifically relating to $tensorflow_root.

Please forgive my amateur performance today, but the location of the root certainly is within the anaconda environment.  What are we looking for?

**Reply by njzjz** [reply]
Sorry but I don't quite understand where you are. Did you use conda to install anything?

For the plugin mode, deepmd-kit and lammps are built separately and individually.

**Reply by npiroozan** [reply]
Right, the two are build separately as one would expect with a plugin mode.

Tensorflow was built from source, as was horovod and other necessary packages.  I got DeePMD running without any problem using Tensorflow 2.9 and Python 3.9 along with horovod 0.24.0.  Then I linked DeePMD to LAMMPS-23June2022 via plugin mode.

Here is a more detailed explanation of where I am:

DeePMD training is complete, with out issue.  I linked DeePMD and Lammps successfully without problem, and ran a full Lammps simulation with DeePMD as the pair_style.  This was done in double precision.

In using Bfloat16 in the neural network, the output is now in single precision, and I need to run LAMMPS in single precision mode as well.  To do this, you specified instructions, but I am missing some of these files such as lmp_low.  My only complied executable within $deepmd_root/source/bin  is lmp.

Thank you!

**Reply by njzjz** [reply]
> Then I linked DeePMD to LAMMPS-23June2022 via plugin mode.

Could you explain it? If you didn't build the C++ interface, I don't think you are able to build load a plugin in the LAMMPS.

Did you follow https://docs.deepmodeling.com/projects/deepmd/en/master/third-party/lammps-command.html#enable-deepmd-kit-plugin-plugin-mode ?

**Reply by npiroozan** [reply]
Certainly.  the way that I did this is to follow the following steps:

https://docs.deepmodeling.com/projects/deepmd/en/master/install/install-lammps.html#install-lammps-plugin-mode

This created an "lmp" executable which I then used to run DeePMD.  I did not follow the steps that you showed in the link above.

**Reply by njzjz** [reply]
I don't think it would work if you didn't enable it.

**Reply by npiroozan** [reply]
I have to say, it did work in DeePMD+LAMMPS.  I'm not sure how it was working, but this is ultimately not the issue.  I would like to now do it the right way by first buliding the C++ interface.  For this I have an error that I detailed earlier, namely:

cmake -DTENSORFLOW_ROOT=/scratch2/npirooza/anaconda3/envs/deepmd29_py309/lib/python3.9/site-packages/tensorflow -DCMAKE_INSTALL_PREFIX=/scratch2/npirooza/Intel_LAMMPS/deepmd-kit-deepmd29_py309 -DLAMMPS_SOURCE_ROOT=/scratch2/npirooza/Intel_LAMMPS/lammps-23Jun2022 ..
-- The C compiler identification is GNU 8.5.0
-- The CXX compiler identification is GNU 8.5.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Could NOT find Git (missing: GIT_EXECUTABLE)
-- Supported model version: 1.1
-- Will not build nv GPU support
-- Will not build AMD GPU support
-- Enabled cpp interface build, looking for tensorflow_cc and tensorflow_framework
CMake Error at cmake/Findtensorflow.cmake:93 (message):
  Not found lib/'tensorflow_cc' in
  '/scratch2/npirooza/anaconda3/envs/deepmd29_py309/lib/python3.9/site-packages/tensorflow;/scratch2/npirooza/anaconda3/envs/deepmd29_py309/lib/python3.9/site-packages/tensorflow/../tensorflow_core;/scratch2/npirooza/anaconda3/envs/deepmd29_py309/lib/python3.9/site-packages/tensorflow;/scratch2/npirooza/anaconda3/envs/deepmd29_py309/lib/python3.9/site-packages/tensorflow/../tensorflow_core;/usr/;/usr/local/'
  You can manually set the tensorflow install path by -DTENSORFLOW_ROOT
Call Stack (most recent call first):
  CMakeLists.txt:78 (find_package)


-- Configuring incomplete, errors occurred!
See also "/scratch2/npirooza/Intel_LAMMPS/deepmd-kit-deepmd29_py309/source/build/CMakeFiles/CMakeOutput.log".

**Reply by npiroozan** [reply]
After this is done we can discuss how my case was working in more detail if that is ok?  I know it may be of interest to code development

**Reply by npiroozan** [reply]
After training, freezing, and compressing the model, I linked DeePMD with LAMMPS in the way I described via the manual instructions.

Then this created the executable lmp.  I went to LAMMPS then and wrote the input script, setting pair_style to DeePMD.

Then I executed this with:

mpirun.......
lmp -in in.silicon

It worked well, no errors at all and gave accurate output.

**Reply by njzjz** [reply]
It's expected to use TensorFlow's C++ interface. It loonks you gave the Python interface.

**Reply by npiroozan** [reply]
Ah ok, indeed, I think you are right; that's what happened.

Do you know how I can resolve the C++ interface issue I'm having?

ThNk you so much.

**Reply by njzjz** [reply]
This is what you gave:
```sh
 -DTENSORFLOW_ROOT=/scratch2/npirooza/anaconda3/envs/deepmd29_py309/lib/python3.9/site-packages/tensorflow
```
It's obviously the Python interface. Did you compile the C++ interface?

**Reply by npiroozan** [reply]
The necessary header files are all within this folder pathway. They include session.h and so forth.  Is something else needed to be done?

**Reply by njzjz** [reply]
If you build from TensorFlow's Python libraries instead of C++ libraries, you have to set `-DUSE_TF_PYTHON_LIBS=TRUE`. Their filenames are different.

**Reply by npiroozan** [reply]
Understood, thank you. I'll try it now.

**Reply by npiroozan** [reply]
I get the same error before after using this flag.  I assure you that the following files are present:

session.h
env.h
op.h
op_kernel.h
shape_inference.h

For clarification, my python version is 3.9 and gcc is 8.5.0

Full command is as follows:

cmake  -DTENSORFLOW_ROOT=/scratch2/npirooza/anaconda3/envs/deepmd29_py309/lib/python3.9/site-packages/tensorflow   
-DCMAKE_INSTALL_PREFIX=/scratch2/npirooza/Intel_LAMMPS/deepmd-kit-deepmd29_py309 
-DLAMMPS_SOURCE_ROOT=/scratch2/npirooza/Intel_LAMMPS/lammps-23Jun2022/src 
-DUSE_TF_PYTHON_LIBS=TRUE ..

Error is:

CMake Error at cmake/Findtensorflow.cmake:93 (message):
  Not found lib/'tensorflow_cc'

**Reply by npiroozan** [reply]
I've tried everything I can, but nothing has worked thus far.

**Reply by njzjz** [reply]
The following files should be present:
- `lib/libtensorflow_framework.so`
- `lib/libtensorflow_cc.so`

I think you can manually check it.

**Reply by npiroozan** [reply]
I do not see it there.  The steps that I took to build tensorflow are as follows:

1.  Create conda environment
2.  Pip Install Python 3.9
3. Pip Install tensorflow 2.9
4. Install DeePMD
5. Set $tensorflow_root and link with DeePMD/source/build

Is there an additional step?

**Reply by njzjz** [reply]
C++ interface is expected as described [here](https://docs.deepmodeling.com/projects/deepmd/en/v2.1.5/install/install-from-source.html#install-the-tensorflow-s-c-interface).

If one wants to use the Python interface, its files must be adjusted to the same as the C++ interface.

**Reply by npiroozan** [reply]
Understood, so positioning of files in directory pathway is the problem it seems?

**Reply by npiroozan** [reply]
Everything is working well now, thank you for your help!
